Damage threshold with von Mises stress calving or eigencalving

trhille opened the pull request:
Allow ice connected to the margin with damage exceeding a user-defined threshold value to calve while also using `config_calving = 'von_Mises_stress'` or `config_calving = 'eigencalving'`. We have found that having both rate- and damage-based calving laws active are necessary to reproduce observations at Thwaites Glacier, for example. To use this feature, set `config_damage_calving_method = 'threshold'` while using `config_calving = 'von_Mises_stress'` or `config_calving = 'eigencalving'`. Unless you are specifying damage as an input field for some reason, you will also want `config_calculate_damage = .true.`.

trhille commented:
Testing shows this behaves as expected. Tested on Thwaites_1to8km_r02_20210914.nc mesh, using the following relevant namelist.landice settings:
```
&velocity_solver
    config_velocity_solver = 'FO'
    config_sia_tangent_slope_calculation = 'from_vertex_barycentric'
    config_flowParamA_calculation = 'PB1982'
    config_do_velocity_reconstruction_for_external_dycore = .false.
    config_simple_velocity_type = 'uniform'
    config_use_glp = .false.
    config_nonconvergence_error = .false.
/
&advection
    config_thickness_advection = 'fo'
    config_tracer_advection = 'fo'
&calving
    config_calving = 'von_Mises_stress'
    config_grounded_von_Mises_threshold_stress = 3.0e5 
    config_floating_von_Mises_threshold_stress = 3.0e5
    config_damage_calving_method = 'threshold'
    config_calculate_damage = .true.
    config_damage_calving_threshold = 0.9
    !config_damage_stiffness_min = 1.0
    config_damage_rheology_coupling = .false.
    config_damage_gl_setting = 'nye'
    config_data_calving = .false.
    config_calving_timescale = 0.0
    config_restore_calving_front = .false.
    config_remove_icebergs = .false.
    config_remove_small_islands = .false.
/
&thermal_solver
    config_thermal_solver = 'none'
    config_thermal_calculate_bmb = .true.
    config_temperature_init = 'file'
    config_thermal_thickness = 0.0
    config_surface_air_temperature_source = 'file'
    config_surface_air_temperature_value = 273.15
    config_surface_air_temperature_lapse_rate = 0.01
    config_basal_heat_flux_source = 'file'
    config_basal_heat_flux_value = 0.0
/
&iceshelf_melt
    config_basal_mass_bal_float = 'temperature_profile'
```

In the following screenshots, damage=0.9 is contoured in cyan.
Start of run:
![image](https://user-images.githubusercontent.com/17446278/161863518-06f65562-39c6-4f88-8ff2-c9848d514927.png)

Just before major damage calving:
![image](https://user-images.githubusercontent.com/17446278/161863617-c040eb66-a8d5-427d-b460-81fc1d890146.png)

And just after:
![image](https://user-images.githubusercontent.com/17446278/161863646-89eeedaf-4f26-4f9b-a661-6314ac2f6a50.png)

End of run:
![image](https://user-images.githubusercontent.com/17446278/161863674-3f1ab74c-d964-4c53-9b56-1d91a61026e1.png)

trhille commented:
@matthewhoffman , one design option I've been going back and forth about is whether we want to add a `'none'` option for `config_damage_calving_method`. This technically would not be any different from `config_damage_calving_method = 'calving_rate` for this usage, but it's more intuitive for a user to know they are turning damage calving off. If we decide to do that, then the default setting for `config_damage_calving_method` should be changed to `none`; otherwise it should be changed to `calving_rate` so that we aren't using damage threshold calving by default when we don't meant to be. I haven't implemented this yet because of these two possible solutions that both have their pros and cons.

matthewhoffman commented:
@trhille , recap of our discussion about this: We want to reorganize the calving options to be less confusing/mistake-prone to a user.  We are ok with deprecating `config_damage_calving_method = 'calving_rate'` if necessary, but it might be less work to keep it in the code if we can do so while keeping the config options easy to understand.

trhille commented:
Force-push from 829ca48 to 6394510 was a rebase onto the head of MALI-Dev/develop (commit 8df83d92551b5dad20d71674ad850ef1b95ed618)

trhille commented:
Passed all COMPASS baseline comparison tests, using @matthewhoffman's baseline directory `/lustre/scratch4/turquoise/mhoffman/COMPASS_REGRESSION/BASELINE`
```
Test Runtimes:
00:12 PASS landice_dome_2000m_sia_restart_test
00:06 PASS landice_dome_2000m_sia_decomposition_test
00:07 PASS landice_dome_variable_resolution_sia_restart_test
00:05 PASS landice_dome_variable_resolution_sia_decomposition_test
00:32 PASS landice_enthalpy_benchmark_A
00:15 PASS landice_eismint2_decomposition_test
00:15 PASS landice_eismint2_enthalpy_decomposition_test
00:15 PASS landice_eismint2_restart_test
00:14 PASS landice_eismint2_enthalpy_restart_test
00:18 PASS landice_greenland_sia_restart_test
00:12 PASS landice_greenland_sia_decomposition_test
00:10 PASS landice_hydro_radial_restart_test
00:06 PASS landice_hydro_radial_decomposition_test
00:16 PASS landice_dome_2000m_fo_decomposition_test
00:15 PASS landice_dome_2000m_fo_restart_test
00:10 PASS landice_dome_variable_resolution_fo_decomposition_test
00:11 PASS landice_dome_variable_resolution_fo_restart_test
00:14 PASS landice_circular_shelf_decomposition_test
00:42 PASS landice_greenland_fo_decomposition_test
00:40 PASS landice_greenland_fo_restart_test
00:33 PASS landice_thwaites_decomposition_test
00:44 PASS landice_thwaites_restart_test
Total runtime 06:33
PASS: All passed successfully!
```

trhille commented:
> @trhille , recap of our discussion about this: We want to reorganize the calving options to be less confusing/mistake-prone to a user. We are ok with deprecating `config_damage_calving_method = 'calving_rate'` if necessary, but it might be less work to keep it in the code if we can do so while keeping the config options easy to understand.

@matthewhoffman this is addressed in https://github.com/MALI-Dev/E3SM/pull/24/commits/4af728c34bcba46267b14d53a2b1231c459fd690, in which I added a `config_damage_calving_method = 'none'` method, added useful log errors, and updated the explanation in Registry.xml.

trhille commented:
Also tested using @matthewhoffman's calving tests in PR https://github.com/MPAS-Dev/compass/pull/318, changing `config_calving_eigencalving_parameter_scalar_value`  from 3.14e16 to 3.14e14 to avoid the >10% ablation error. We've seen the failure on the eigencalving restart test before, in which the restart run encounters the >10% ablation error while the full run does not. That needs to be fixed, but is not unique to this PR. The ismip6_retreat test fails on validation, also not related to this PR.

```
Test Runtimes:
01:04 PASS landice_humboldt_mesh-default_decomposition_test_velo-none_calving-none
01:02 PASS landice_humboldt_mesh-default_decomposition_test_velo-none_calving-floating
01:03 PASS landice_humboldt_mesh-default_decomposition_test_velo-none_calving-eigencalving
01:03 PASS landice_humboldt_mesh-default_decomposition_test_velo-none_calving-specified_calving_velocity
01:03 PASS landice_humboldt_mesh-default_decomposition_test_velo-none_calving-von_mises_stress
01:02 PASS landice_humboldt_mesh-default_decomposition_test_velo-none_calving-damagecalving
01:04 FAIL landice_humboldt_mesh-default_decomposition_test_velo-none_calving-ismip6_retreat
00:22 PASS landice_humboldt_mesh-default_restart_test_velo-none_calving-none
00:22 PASS landice_humboldt_mesh-default_restart_test_velo-none_calving-floating
00:20 FAIL landice_humboldt_mesh-default_restart_test_velo-none_calving-eigencalving
00:22 PASS landice_humboldt_mesh-default_restart_test_velo-none_calving-specified_calving_velocity
00:22 PASS landice_humboldt_mesh-default_restart_test_velo-none_calving-von_mises_stress
00:22 PASS landice_humboldt_mesh-default_restart_test_velo-none_calving-damagecalving
00:23 PASS landice_humboldt_mesh-default_restart_test_velo-none_calving-ismip6_retreat
Total runtime 09:55
FAIL: 2 tests failed, see above.
```

matthewhoffman commented:
Those changes look good and it looks like you did thorough testing.  I think your solution of how to handle the combining of the options is good, all things considered, and the explanation in the Registry option for `config_damage_calving_method` is clear.  One picky question - should we also add a note in the `config_calving` option saying threshold damage calving can be combined with vM & eigen and to see the `config_damage_calving_method` for details?
